OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs

```diff
@@ -135,26 +108,16 @@ public override void DoImpact(Target target, WarheadArgs args)
 				Game.Sound.Play(SoundType.World, impactSound, pos);
 		}
 
-		public bool IsValidImpact(WPos pos, Actor firedBy)
+		/// <summary>Checks if the warhead is valid against the terrain at impact position.</summary>
+		bool IsValidAgainstTerrain(World world, WPos pos)
 		{
-			var world = firedBy.World;
-			var targetTile = world.Map.CellContaining(pos);
-			if (!world.Map.Contains(targetTile))
+			var cell = world.Map.CellContaining(pos);
+			if (!world.Map.Contains(cell))
 				return false;
 
-			var impactType = GetImpactType(world, targetTile, pos, firedBy);
-			switch (impactType)
-			{
-				case ImpactType.TargetHit:
-					return true;
-				case ImpactType.Air:
-					return IsValidTarget(TargetTypeAir);
-				case ImpactType.Ground:
-					var tileInfo = world.Map.GetTerrainInfo(targetTile);
-					return IsValidTarget(tileInfo.TargetTypes);
-				default:
-					return false;
-			}
+			var dat = world.Map.DistanceAboveTerrain(pos);
+			var tileInfo = world.Map.GetTerrainInfo(cell);
+			return IsValidTarget(dat > AirThreshold ? TargetTypeAir : tileInfo.TargetTypes);
```

Comment:
Fwiw, the tile info lookup is relatively expensive, so it might make sense to inline the variable here.
`return IsValidTarget(dat > AirThreshold ? TargetTypeAir : world.Map.GetTerrainInfo(cell).TargetTypes);`